controller/rancher/rancher.go

```diff
@@ -488,6 +550,238 @@ func (fetcher *RCertificateFetcher) UpdateEndpoints(lbSvc *metadata.Service, eps
+		//read from cache if certDir matches and certs are cached.
+		logrus.Debug("Reading certs from cache")
+	} else {
+		// init/re-init cache
+		fetcher.mu.Lock()
+		fetcher.removeWatches()
+
+		fetcher.CertsCache = make(map[string]*config.Certificate)
+		logrus.Debug("Reading certs from directory and starting watchers")
+		fetcher.TempCertsMap = make(map[string]*config.Certificate)
+		err := filepath.Walk(certDir, fetcher.readCertificate)
+		if err != nil {
+			fetcher.CertDir = ""
+			fetcher.CertsCache = nil
+			logrus.Errorf("Error %v reading certs from directory %v", err, certDir)
+		} else {
+			fetcher.CertDir = certDir
+			for path, cert := range fetcher.TempCertsMap {
+				fetcher.CertsCache[path] = cert
+			}
+			fetcher.addWatches()
+		}
+		fetcher.mu.Unlock()
+	}
+
+	fetcher.mu.RLock()
+	for _, value := range fetcher.CertsCache {
+		certs = append(certs, value)
+	}
+	fetcher.mu.RUnlock()
+
+	return certs, nil
+}
+
+func (fetcher *RCertificateFetcher) ReadDefaultCertificate(defaultCertDir string) (*config.Certificate, error) {
+	fetcher.mu.RLock()
+	currentDefCertD := fetcher.DefaultCertDir
+	fetcher.mu.RUnlock()
+
+	if currentDefCertD != defaultCertDir {
+		//defaultCertDir value changed, read it again.
+		fetcher.Watcher.Remove(defaultCertDir)
+		fetcher.mu.Lock()
+		fetcher.TempCertsMap = make(map[string]*config.Certificate)
+		err := filepath.Walk(defaultCertDir, fetcher.readCertificate)
+		if err != nil {
+			fetcher.DefaultCertDir = ""
+			fetcher.DefaultCert = nil
+			logrus.Errorf("Error %v reading defaultCertDir  %v", err, defaultCertDir)
+		} else {
+			fetcher.DefaultCertDir = defaultCertDir
+			for _, cert := range fetcher.TempCertsMap {
+				fetcher.DefaultCert = cert
+			}
+			fetcher.Watcher.Add(defaultCertDir)
+		}
+		fetcher.mu.Unlock()
+	}
+	return fetcher.DefaultCert, nil
+}
+
+func (fetcher *RCertificateFetcher) readCertificate(path string, f os.FileInfo, err error) error {
+	if f != nil && f.IsDir() {
+		logrus.Debugf("walking %v", path)
+		isCertFound := false
+		cert := config.Certificate{}
+		cert.Name = f.Name()
+		files, err := ioutil.ReadDir(path)
+		if err != nil {
+			return err
+		}
+		for _, file := range files {
+			if !file.IsDir() {
+				contentBytes, err := fetcher.evaluatueLinkAndReadFile(path, file.Name())
+				if err != nil {
+					return err
+				}
+				if file.Name() == "fullchain.pem" {
+					isCertFound = true
+					cert.Cert = string(*contentBytes)
+				} else if file.Name() == "privkey.pem" {
+					isCertFound = true
+					cert.Key = string(*contentBytes)
+				}
+			}
+		}
+		if isCertFound {
+			fetcher.TempCertsMap[path] = &cert
+		}
+	}
+	return nil
+}
+
+func (fetcher *RCertificateFetcher) evaluatueLinkAndReadFile(relativePath string, fileName string) (*[]byte, error) {
+	filePath := path.Join(relativePath, fileName)
+
+	absFilePath, err := filepath.Abs(filePath)
+	if err != nil {
+		logrus.Errorf("Error forming path to file %s, error: %v", filePath, err)
```

Comment:
we should not log errors here as they are already being logged in the caller stack. Just remove line 669 and do "return nil, fmt.Errorf()" construct on line 670. Apply to any other place in the code that has an unnecessary logging (line 675, may be more...)